idx/bigtable/bigtable.go

```diff
@@ -259,56 +259,63 @@ func (b *BigtableIdx) AddOrUpdate(mkey schema.MKey, data *schema.MetricData, par
 				if err != nil {
-					log.Errorf("bigtable-idx: Failed to delete row %s: %s", archive.Id, err)
+					log.Errorf("bigtable-idx: Failed to delete row %s: %s", id, err)
 				}
 			}()
 		}
 	}
 
 	// check if we need to save to bigtable.
 	now := uint32(time.Now().Unix())
-	if archive.LastSave < (now - b.cfg.updateInterval32) {
-		archive = b.updateBigtable(now, inMemory, archive, partition)
+	if lastSave < (now - b.cfg.updateInterval32) {
+		b.updateBigtable(now, inMemory, id, partition, lastSave)
 	}
 
 	stat.Value(time.Since(pre))
-	return archive, oldPartition, inMemory
+	return id, lastSave, oldPartition, inMemory
 }
 
 // updateBigtable saves the archive to bigtable and
 // updates the memory index with the updated fields.
-func (b *BigtableIdx) updateBigtable(now uint32, inMemory bool, archive *interning.ArchiveInterned, partition int32) *interning.ArchiveInterned {
+func (b *BigtableIdx) updateBigtable(now uint32, inMemory bool, id schema.MKey, partition int32, lastSave uint32) {
 	// if the entry has not been saved for 1.5x updateInterval
 	// then perform a blocking save.
-	if archive.LastSave < (now - b.cfg.updateInterval32 - (b.cfg.updateInterval32 / 2)) {
+	if lastSave < (now - b.cfg.updateInterval32 - (b.cfg.updateInterval32 / 2)) {
 		if log.IsLevelEnabled(log.DebugLevel) {
-			log.Debugf("bigtable-idx: updating def %s in index.", archive.Id)
+			log.Debugf("bigtable-idx: updating def %s in index.", id)
+		}
+		arc, ok := b.MemoryIndex.Get(id)
+		if ok {
+			b.writeQueue <- writeReq{recvTime: time.Now(), def: arc.CloneInterned()}
+			b.MemoryIndex.UpdateArchiveLastSave(id, partition, now)
 		}
-		b.writeQueue <- writeReq{recvTime: time.Now(), def: archive.CloneInterned()}
-		archive.LastSave = now
-		b.MemoryIndex.UpdateArchiveLastSave(archive.Id, archive.Partition, now)
 	} else {
 		// perform a non-blocking write to the writeQueue. If the queue is full, then
 		// this will fail and we won't update the LastSave timestamp. The next time
 		// the metric is seen, the previous lastSave timestamp will still be in place and so
 		// we will try and save again.  This will continue until we are successful or the
 		// lastSave timestamp become more then 1.5 x UpdateInterval, in which case we will
 		// do a blocking write to the queue.
-		select {
-		case b.writeQueue <- writeReq{recvTime: time.Now(), def: archive.CloneInterned()}:
-			archive.LastSave = now
-			b.MemoryIndex.UpdateArchiveLastSave(archive.Id, archive.Partition, now)
-		default:
-			statSaveSkipped.Inc()
-			if log.IsLevelEnabled(log.DebugLevel) {
-				log.Debugf("bigtable-idx: writeQueue is full, update of %s not saved this time", archive.Id)
+		arc, ok := b.MemoryIndex.Get(id)
+		if ok {
+			select {
+			// note: this attempt to write actually creates the struct, which calls
+			// Now() and CloneInterned(), even if the write fails
+			case b.writeQueue <- writeReq{recvTime: time.Now(), def: arc.CloneInterned()}:
```

Comment:
same here, this `.CloneInterned()` isn't necessary, then the `arch.ReleaseInterned()` further down can also be removed. important is just that the properties of `arc` don't get accessed inside this go-routine after it's been pushed into the chan